user-guide: add troubleshooting entry for dvc-exp in shallow clones (#3577)
iterative/dvc.org: content/docs/user-guide/troubleshooting.md

[internal directories]:
  https://dvc.org/doc/user-guide/project-structure/internal-files

## DVC experiments may fail in shallow Git repositories {#git-shallow}

> DVC experiments rely on Git features which may not work properly in shallow Git

Reviewer: Maybe include a link to https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---depthltdepthgt or elsewhere to explain shallow clones?

> repositories. This error commonly occurs in certain CI environments which use

Reviewer: Not sure everyone will know what we mean by "CI environments" or even what CI is. Maybe at least we can mention some of those common environments, like GitHub Actions and GitLab CI?

> DVC experiments rely on Git features which may not work properly in shallow Git
> repositories. This error commonly occurs in certain CI environments which use
> shallow git clones by default. When this error is encountered, try unshallowing
> the Git repository and then retry the DVC command.
> 
> Local Git repositories can be unshallowed by using:
> 
> ```
> $ git fetch --unshallow
> ```

Reviewer 2: ```suggestion
`dvc exp` commands use internal Git operations which may not work properly in
[shallow clones](https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---depthltdepthgt).
Local Git repositories can be unshallowed with `git fetch --unshallow`.

Typically, however, this occurs in transient remote environments such as
Continuous Integration jobs, which use shallow clones by default. In those cases,
change their configuration to avoid shallow cloning. Common examples:
```

When using [CML](https://cml.dev/doc), repositories can be unshallowed by using:

```
$ cml ci --unshallow
```

> In GitHub Actions, repositories can be unshallowed by setting `fetch-depth` in

Reviewer: Can we further clarify that users should either do the CML option or one of the GH/GL options?

Reviewer 2: Same.

> When using [CML](https://cml.dev/doc), repositories can be unshallowed by using:
> 
> ```
> $ cml ci --unshallow
> ```
> 
> In GitHub Actions, repositories can be unshallowed by setting `fetch-depth` in
> the `actions/checkout` action.
> 
> ```
> - uses: actions/checkout@v3
>   with:
>     fetch-depth: 0
> ```
> 
> In GitLab CI/CD, repositories can be unshallowed by setting the `GIT_DEPTH`
> environment variable:
> 
> ```
> variables:
>   GIT_DEPTH: "0"
> ```

Reviewer 2: ````suggestion
<toggle>
<tab title="CML">

When using [CML](https://cml.dev/doc), repositories can be unshallowed by using:

```cli
$ cml ci --unshallow
```

</tab>
<tab title="GitHub Actions">

Set `fetch-depth` to `0` in the `actions/checkout` action:

```yaml
- uses: actions/checkout@v3
  with:
    fetch-depth: 0
```

> More info.
> [here](https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches).

</tab>
<tab title="GitLab CI/CD">

Set the `GIT_DEPTH` env var to `0`:

```yaml
variables:
  GIT_DEPTH: "0"
```

> More info.
> [here](https://docs.gitlab.com/ee/ci/large_repositories/#shallow-cloning).

</tab>
</toggle>
````

[internal directories]:
  https://dvc.org/doc/user-guide/project-structure/internal-files

## DVC Experiments may fail in Git shallow clones {#git-shallow}

`dvc exp` commands use internal Git operations which may not work properly in
[shallow clones](https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---depthltdepthgt).
Local Git repositories can be unshallowed with `git fetch --unshallow`.

Typically, however, this occurs in transient remote environments such as
Continuous Integration jobs, which use shallow clones by default. In those
cases, change their configuration to avoid shallow cloning. Common examples:

<toggle>
<tab title="CML">

> When using [CML](https://cml.dev/doc), repositories can be unshallowed by using:

Reviewer 2: Copy edit
```suggestion
[CML](https://cml.dev/doc) has a convenient `--unshallow` option for it's
[`ci`](https://cml.dev/doc/ref/ci) command:
```

> ```cli
> $ cml ci --unshallow
> ```

Reviewer 2: This ain't working on this repo yet...

https://dvcorg-troubleshootingshallow.gtsb.io/doc/user-guide/troubleshooting?tab=CML#git-shallow

````suggestion
```dvc
$ cml ci --unshallow
```
````

Reviewer 3: resolving since `cli` should work now afaik

[internal directories]:
  https://dvc.org/doc/user-guide/project-structure/internal-files

## DVC Experiments may fail in Git shallow clones {#git-shallow}

> `dvc exp` commands use internal Git operations which may not work properly in

Reviewer 3: maybe hyperlink to more info (presumably some section on reflog things?)

Reviewer: Could use https://dvc.org/doc/user-guide/experiment-management/experiments-overview#how-does-dvc-track-experiments or https://dvc.org/blog/experiment-refs.

Reviewer 2: We don't explain internal git merges anywhere, not even in Peter's blog post. Too deep

[shallow clones](https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---depthltdepthgt).
Local Git repositories can be unshallowed with `git fetch --unshallow`.

> Typically, however, this occurs in transient remote environments such as

Reviewer 3: ```suggestion
This often occurs in transient remote environments such as
```

> Continuous Integration jobs, which use shallow clones by default. In those

Reviewer 3: ```suggestion
Continuous Integration (CI) jobs, which use shallow clones by default. In those
```

cases, change their configuration to avoid shallow cloning. Common examples:

<toggle>
<tab title="CML">

> [CML](https://cml.dev/doc) has a convenient `--unshallow` option for it's
> [`ci`](https://cml.dev/doc/ref/ci) command:

Reviewer 3: ```suggestion
[CML](https://cml.dev) has a convenient `--unshallow` option for its
[`ci`](https://cml.dev/doc/ref/ci) command:
```

```cli
$ cml ci --unshallow
```

</tab>
<tab title="GitHub Actions">

Set `fetch-depth` to `0` in the `actions/checkout` action:

```yaml
- uses: actions/checkout@v3
  with:
    fetch-depth: 0
```

> > More info.
> > [here](https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches).

Reviewer 3: is this meant to be a callout?

```suggestion
<admon type="info">

See the [GitHub Actions docs](https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches) for more information.

</admon>
```

</tab>
<tab title="GitLab CI/CD">

Set the `GIT_DEPTH` env var to `0`:

```yaml
variables:
  GIT_DEPTH: '0'
```

> > More info.
> > [here](https://docs.gitlab.com/ee/ci/large_repositories/#shallow-cloning).

Reviewer 3: is this meant to be a callout?

```suggestion
<admon type="info">

See the [GitLab CI/CD docs](https://docs.gitlab.com/ee/ci/large_repositories/#shallow-cloning) for more information.

</admon>
```
